### PR TITLE
[minor_changes] Added additional arguments to schema_template_vrf resource.

### PR DESCRIPTION
**Pull request by anvitha-jain**
Added ip_data_plane_learning and preferred_group arguments to schema_template_vrf resource

### CODE REVIEW

File: plugins/modules/mso_schema_template_vrf.py [hunk 1]
```diff
@@ -46,6 +47,16 @@
     description:
     - Whether to enable vzAny.
     type: bool
+  ip_data_plane_learning:
+    description:
+    - Type of IP Data-Plane Learning configuration. 
+    type: str
+    choices: [ disabled, enabled ]
+    default: enabled
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_vrf.py:55]: is this needed?

File: plugins/modules/mso_schema_template_vrf.py [hunk 2]
```diff
@@ -46,6 +47,16 @@
     description:
     - Whether to enable vzAny.
     type: bool
+  ip_data_plane_learning:
+    description:
+    - Type of IP Data-Plane Learning configuration.
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_vrf.py:52]: Not really a type. I would rephrase

**Review by lhercot — CHANGES_REQUESTED** [review]

File: plugins/modules/mso_schema_template_vrf.py (hunk 1 above)
Comment (anvitha-jain) [reply plugins/modules/mso_schema_template_vrf.py]: removed

File: plugins/modules/mso_schema_template_vrf.py [hunk 3]
```diff
@@ -46,6 +47,15 @@
     description:
     - Whether to enable vzAny.
     type: bool
+  ip_data_plane_learning:
```
Comment (akinross) [inline plugins/modules/mso_schema_template_vrf.py:50]: should we use boolean type, and then map to enabled/disabled?

File: plugins/modules/mso_schema_template_vrf.py [hunk 4]
```diff
@@ -46,6 +47,15 @@
...
+    description:
+    - Whether IP data plane learning is enabled or disabled.
+    type: str
```
Comment (akinross) [inline plugins/modules/mso_schema_template_vrf.py:53]: should we use boolean type, and then map to enabled/disabled?

File: plugins/modules/mso_schema_template_vrf.py [hunk 5]
```diff
@@ -46,6 +47,15 @@
     description:
     - Whether to enable vzAny.
     type: bool
+  ip_data_plane_learning:
+    description:
+    - Whether IP data plane learning is enabled or disabled.
```
Comment (akinross) [inline plugins/modules/mso_schema_template_vrf.py:52]: Is there a default when not provided? If so should this be added to the description?

File: tests/integration/targets/mso_schema_template_vrf/tasks/main.yml [hunk 6]
```diff
@@ -146,6 +147,7 @@
     vrf: VRF2
     layer3_multicast: true
     vzany: true
+    ip_data_plane_learning: disabled
```
Comment (akinross) [inline tests/integration/targets/mso_schema_template_vrf/tasks/main.yml:150]: Should we add test for enabled case?

**Review by akinross — CHANGES_REQUESTED** [review]

File: plugins/modules/mso_schema_template_vrf.py (hunk 5 above)
Comment (anvitha-jain) [reply plugins/modules/mso_schema_template_vrf.py:52]: The default value set when you don't send it. which is why we do not need default.
https://github.com/CiscoDevNet/ansible-mso/pull/358#discussion_r1157651052

File: tests/integration/targets/mso_schema_template_vrf/tasks/main.yml (hunk 6 above)
Comment (anvitha-jain) [reply tests/integration/targets/mso_schema_template_vrf/tasks/main.yml:150]: the value is enabled by default. to validate added an assert in name: Verify vrf3_nm

File: tests/integration/targets/mso_schema_template_vrf/tasks/main.yml [hunk 7]
```diff
@@ -515,4 +519,55 @@
...
+    - nm_query_change_vrf_contract_again.current | length == 2
+
+#  Verifying preferred Endpoint Group for VRF.
+- name: Add VRF preferred group (normal mode)
+  mso_schema_template_vrf:
+    <<: *mso_info
+    schema: '{{ mso_schema | default("ansible_test") }}'
+    template: Template 2
+    vrf: vrf_preferred_group
+    preferred_group: true
```
[the remark this answers is not in the document]
Comment (anvitha-jain) [reply tests/integration/targets/mso_schema_template_vrf/tasks/main.yml:531]: the value is false by default. to validate added an assert in name: Verify vrf3_nm

File: plugins/modules/mso_schema_template_vrf.py [hunk 8]
```diff
@@ -46,6 +47,15 @@
...
+    type: str
+    choices: [ disabled, enabled ]
+  preferred_group:
+    description:
+    - Whether to enable preferred Endpoint Group.
```
[the remark this answers is not in the document]
Comment (anvitha-jain) [reply plugins/modules/mso_schema_template_vrf.py:57]: The default value set when you don't send it. which is why we do not need default.
https://github.com/CiscoDevNet/ansible-mso/pull/358#discussion_r1157651052

File: plugins/modules/mso_schema_template_vrf.py (hunk 4 above)
Comment (anvitha-jain) [reply plugins/modules/mso_schema_template_vrf.py:53]: will keep it as string for now.

File: plugins/modules/mso_schema_template_vrf.py (hunk 5 above)
Comment (akinross) [reply plugins/modules/mso_schema_template_vrf.py:53]: Correct this is what I thought, but then we document in the description.
```suggestion
    - Whether IP data plane learning is enabled or disabled.
    - The APIC defaults to C(enabled) when unset during creation.
```

File: plugins/modules/mso_schema_template_vrf.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (akinross) [reply plugins/modules/mso_schema_template_vrf.py:58]:
```suggestion
    - Whether to enable preferred Endpoint Group.
    - The APIC defaults to C(false) when unset during creation.
```

**Review by akinross — CHANGES_REQUESTED** [review]

**Review by akinross — DISMISSED** [review]

File: tests/integration/targets/mso_schema_template_vrf/tasks/main.yml [hunk 9]
```diff
@@ -515,4 +521,55 @@
...
+
+- name: Verify vrf_preferred_group_nm
+  assert:
+    that:
+    - vrf_preferred_group_nm is changed
+    - vrf_preferred_group_nm.current.name == 'vrf_preferred_group'
+    - vrf_preferred_group_nm.current.displayName == 'vrf_preferred_group'
+    - vrf_preferred_group_nm.current.preferredGroup == True
+
+# ADD BD
```
Comment (shrsr) [inline tests/integration/targets/mso_schema_template_vrf/tasks/main.yml:545]: Are the following tasks added just for code coverage?

File: tests/integration/targets/mso_schema_template_vrf/tasks/main.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (shrsr) [reply tests/integration/targets/mso_schema_template_vrf/tasks/main.yml:533]:
```
- name: Add VRF3 (normal mode)
   mso_schema_template_vrf:
    <<: *mso_info
    schema: '{{ mso_schema | default("ansible_test") }}_2'
    template: Template 3
    vrf: VRF3
    state: present
  register: vrf3_nm
  
  - name: Add VRF3 with pref group (normal mode)
  mso_schema_template_vrf:
    <<: *mso_info
    schema: '{{ mso_schema | default("ansible_test") }}_2'
    template: Template 3
    vrf: VRF3
    preferred_group: true
    state: present
  register: vrf3_nm2
  ```
  Does vrf3_nm2 change in the above case?

**Review by shrsr — CHANGES_REQUESTED** [review]

File: tests/integration/targets/mso_schema_template_vrf/tasks/main.yml (hunk 9 above)
Comment (anvitha-jain) [reply tests/integration/targets/mso_schema_template_vrf/tasks/main.yml:545]: Yes and no, the below tasks shows a particular scenario - you can select the preferred group option in the VRF but to enable it you should also make changes in the EPG.

File: tests/integration/targets/mso_schema_template_vrf/tasks/main.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (anvitha-jain) [reply tests/integration/targets/mso_schema_template_vrf/tasks/main.yml:533]: updates it.

File: tests/integration/targets/mso_schema_template_vrf/tasks/main.yml (hunk 9 above)
Comment (shrsr) [reply tests/integration/targets/mso_schema_template_vrf/tasks/main.yml:545]: Since this is not an example, it would be good to verify those with assertions

File: tests/integration/targets/mso_schema_template_vrf/tasks/main.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (shrsr) [reply tests/integration/targets/mso_schema_template_vrf/tasks/main.yml:533]: Can we add a test case for the above scenario?

File: plugins/modules/mso_schema_template_vrf.py (hunk 5 above)
Comment (shrsr) [reply plugins/modules/mso_schema_template_vrf.py:53]: Whether IP data plane learning is enabled or disabled. -> 
Allows setting IP data plane learning as enabled or disabled

File: plugins/modules/mso_schema_template_vrf.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (shrsr) [reply plugins/modules/mso_schema_template_vrf.py:59]: Whether to enable preferred Endpoint Group -> Allows enabling the preferred Endpoint Group

**Review by shrsr — CHANGES_REQUESTED** [review]

File: plugins/modules/mso_schema_template_vrf.py (hunk 5 above)
Comment (anvitha-jain) [reply plugins/modules/mso_schema_template_vrf.py:53]: Since all the previous arguments are explained that way, changing explanation for one argument is not feasible.

File: plugins/modules/mso_schema_template_vrf.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (anvitha-jain) [reply plugins/modules/mso_schema_template_vrf.py:59]: Since all the previous arguments are explained that way, changing explanation for one argument is not feasible.

File: tests/integration/targets/mso_schema_template_vrf/tasks/main.yml (hunk 9 above)
Comment (anvitha-jain) [reply tests/integration/targets/mso_schema_template_vrf/tasks/main.yml:545]: removed this part as discussed

File: tests/integration/targets/mso_schema_template_vrf/tasks/main.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (anvitha-jain) [reply tests/integration/targets/mso_schema_template_vrf/tasks/main.yml:533]: Added asserts in change vrf2 section

**Review by akinross — APPROVED** [review]

**Review by akinross — APPROVED** [review]
LGTM

**Review by lhercot — APPROVED** [review]
LGTM
